### PR TITLE
Split BPFMap API into managed and unmanaged

### PR DESCRIPTION
**Pull request by geyslan**
Close: #354

commit 9cb8ba98615d0dc9f774abd2528227ca6ad33f4b (HEAD -> map-effort, origin/map-effort)
Author: Geyslan Gregório <geyslan@gmail.com>
Date:   Mon Aug 7 17:38:03 2023 -0300

    feat(map): introduce ReuseFD() to reuse a map fd
    
    The `ReuseFD` function links the current BPFMap instance to a provided
    map file descriptor. This enables the reuse of a map that was initiated
    by another process.
    
    Initial effort: #346
    
    Co-authored-by: Francisco Javier Honduvilla Coto <javierhonduco@gmail.com>

commit fe71839b6554ac7e4b97f74f9763897dc4244abf
Author: Geyslan Gregório <geyslan@gmail.com>
Date:   Wed Aug 9 16:46:41 2023 -0300

    chore(map): logic refactoring and standardization
    
    Introduced struct helpers to manage the C struct lifecycle on the C side,
    which avoids problems with structs which may contain bitfields. See #244.
    
      These are replacements for 'BPFMapBatchOpts' and 'bpfMapBatchOptsToC()'.
      - cgo_bpf_map_batch_opts_new
      - cgo_bpf_map_batch_opts_free
    
      These are replacements for 'bpfMapCreateOptsToC()'
      - cgo_bpf_map_create_opts_new
      - cgo_bpf_map_create_opts_free
    
    Restructured libbpf BPF map logic into separate files:
    - 'map-common.go' contains generic BPF map types, constants, and helpers
      like 'GetMapInfoByFD()'.
    - 'map-low.go' includes 'BPFMapLow' and respective logic and helpers
      like 'CreateMap()'.
    - 'map-iterator.go' contains related types and logic.
    
    In the 'BPFMap' struct:
    - Removed fields like 'fd' and 'name' since they must be retrieved
      directly from 'libbpf'.
    - Added 'bpfMapLow' instance for access to low-level methods.
    - Deprecated 'BPFMap.GetMaxEntries' in favor of using 'MaxEntries'.
    - Exposed 'InitialValue' and 'SetInitialValue' as public wrappers.
    - Introduced MapFlags(), IfIndex() and MapExtra() methods.
    
    Introduced 'misc.go' as a place for miscellaneous generic helpers.
    
    Bug Fixes:
    - Fixed cases where 'BPFMap.ValueSize()', possibly 0, was being passed
      directly to 'make()' - detected on the map of maps effort #343.
    
    Further Refactoring and Standardization on the map related logic and on
    lines touched by other changes:
    - Removed check for 'BPFMap.Name()' against nil, as 'C.GoString(C.NULL)'
      returns "".
    - Changed error returns to 'retC' in functions that return error values
      to avoid confusion with possible use of errno.
    - Changed returns to 'valueC' in functions that return values or
      pointers to values.
    - Suffixed variable names that are C types with 'C'.
    - Applied the use of 'defer' for better resource management.

commit 7b73fc0008db9ca42c041cab0f05dbbf18b6aab7
Author: Geyslan Gregório <geyslan@gmail.com>
Date:   Wed Aug 9 15:42:49 2023 -0300

    fix(c): check against NULL before memcpy

commit 8b1190ade69ec1de5073bd5cdb797a6e5de401a5
Author: Geyslan Gregório <geyslan@gmail.com>
Date:   Wed Aug 9 14:17:13 2023 -0300

    chore(c): prefix helpers with 'cgo_'
    
    Prefix all C helpers used in Go with 'cgo_', renaming two cases
    where the functions does not reflect the libbpf API:
    
    set_print_fn -> cgo_libbpf_set_print_fn
    get_internal_map_init_value -> cgo_bpf_map__initial_value

commit f58df61144d79d0621362a9252f0fad817f40f64
Author: Geyslan Gregório <geyslan@gmail.com>
Date:   Wed Aug 9 14:09:17 2023 -0300

    chore(c): move C definitions to 'libbpfgo.c'
    
    Leaving 'libbpfgo.h' only with declarations to avoid multiple
    declaration errors when importing 'libbpfgo.h' from different Go files.

commit c9b7e751bc93baf489d11979f480698ece662261
Author: Geyslan Gregório <geyslan@gmail.com>
Date:   Wed Aug 9 13:53:53 2023 -0300

    chore(map): relocate map logic to map.go

commit d45e8f61e9c60effd666abc5eddea971c207e587
Author: Geyslan Gregório <geyslan@gmail.com>
Date:   Tue Aug 8 16:12:26 2023 -0300

    chore: del unneeded assignment to blank identifier

### CODE REVIEW

**Comment by javierhonduco** [comment]
Thanks for doing this!

**Comment by geyslan** [comment]
Since this involved creating new files and relocating logic, it required making major changes in a single commit.

**Comment by geyslan** [comment]
When this get merged, the next to come to light is the map of maps. Despite the discussion in https://github.com/aquasecurity/libbpfgo/pull/343#issuecomment-1660998211, I'm not so sure about creating another type like `BPFMapOfMaps`, as that would require another module method like `GetMapOfMaps()`. Perhaps best is to leave `BPFMap` and `BPFMapLow` with the same libbpf API and make usage assumptions based on respective selftest. Let's see.

**Comment by geyslan** [comment]
> Since this involved creating new files and relocating logic, it required making major changes in a single commit.

I'm going redo the bigger commit splitting it by subject.

**Comment by geyslan** [comment]
> I'm going redo the bigger commit splitting it by subject for better understanding and review.

@yanivagman I've broken it down a bit more and left `GetMapByID()` and `GetMapFDByID()` implementations for an upcoming PR (map of maps). `ReuseFD()`, co-authored by @javierhonduco, is coming with this since it's a good example of map high and low level separation.

File: rwArray_test.go [hunk 1]
```diff
@@ -109,7 +109,7 @@ func TestRWArrayConcurrent(t *testing.T) {
 
 			// time to exit?
 			select {
-			case _, _ = <-stop:
+			case <-stop:
```
Comment (geyslan) [inline rwArray_test.go:112]: Just a cleanup addressing vscode alerts.

File: map-low.go [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+//
+// BPFMapLow (low-level API)
+//
+
+// BPFMapLow provides a low-level interface to BPF maps.
+// Its methods follow the BPFMap naming convention.
+type BPFMapLow struct {
+	fd   int
+	info *BPFMapInfo
+}
```
[the remark this answers is not in the document]
Comment (javierhonduco) [reply map-low.go:24]: `BpfMapInner`? `BpfMapDetail`?

File: libbpfgo.go [hunk 3]
```diff
@@ -475,491 +389,71 @@ func (m *Module) InitGlobalVariable(name string, value interface{}) error {
...
+				Name:                  bpfMap.Name(),
+				IfIndex:               bpfMap.IfIndex(),
+				BTFVmlinuxValueTypeID: 0,
+				NetnsDev:              0,
+				NetnsIno:              0,
+				BTFID:                 0,
+				BTFKeyTypeID:          0,
+				BTFValueTypeID:        0,
+				MapExtra:              bpfMap.MapExtra(),
+			},
```
Comment (geyslan) [inline libbpfgo.go:445]: This error was detected while implementing the policies version on Tracee running kernel 5.15.75. It doesn't occur in 6.1.44: https://elixir.bootlin.com/linux/v6.1.44/source/tools/lib/bpf/gen_loader.c#L470

File: selftest/reuse-fd/main.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		log.Fatal(err)
+	}
+
+	val2, err := toReuseCreated.GetValue(key2Unsafe) // lookup "tester"
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if bytes.Equal(val2, value2) {
+		log.Fatal("wrong value")
```
Comment (yanivagman) [inline selftest/reuse-fd/main.go:89]: Shouldn't val2 and value2 be equal (since both are taken from tester map?)

**Review by yanivagman — COMMENTED** [review]
Overall LGTM, 
The last commit (ReuseFD) should be feat and not chore,
Also two small questions in the following comments

File: libbpfgo.c [hunk 5]
```diff
@@ -38,12 +38,12 @@ int libbpf_print_fn(enum libbpf_print_level level, // libbpf print level
     return ret;
 }
 
-void set_print_fn()
+void libbpfgo_libbpf_set_print_fn()
```
[the remark this answers is not in the document]
Comment (geyslan) [reply libbpfgo.c:41]: `libbpfgo_` denotes functions C libbpfgo. An alternative is prefixing it with `c_` instead. WDYT?

[the remark this answers is not in the document]
Comment (yanivagman) [reply libbpfgo.c:41]: I see.
So maybe use `cgo` prefix?

File: selftest/reuse-fd/main.go (hunk 4 above)
Comment (geyslan) [reply selftest/reuse-fd/main.go:89]: Yes. Good catch. It is a leftover from a previous test that passed through. I'm fixing this test. Thanks.

**Comment by geyslan** [comment]
> Overall LGTM, The last commit (ReuseFD) should be feat and not chore, Also two small questions in the following comments

Thanks for reviewing this. I'm going to address all requests and bugs.

File: map-low.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (geyslan) [reply map-low.go:24]: Let's stick with BPFMapLow. 👍🏼

**Pull request merged**